app/objects/c_operation.py

```diff
@@ -187,10 +187,9 @@ def update_untrusted_agents(self, agent):
 
     async def all_facts(self):
         knowledge_svc_handle = BaseService.get_service('knowledge_svc')
-        data_svc_handle = BaseService.get_service('data_svc')
         seeded_facts = []
         if self.source:
-            seeded_facts = await data_svc_handle.get_facts_from_source(self.source.id)
+            seeded_facts = await knowledge_svc_handle.get_facts(criteria=dict(source=self.source.id))
```

Comment:
It sounds like these changes from https://github.com/mitre/caldera/pull/2978 were causing issues in your tests - I'd like to avoid undoing changes from previous PRs and fix the corresponding tests if possible. 

@elegantmoose @clenk thoughts?

Reply:
Correct, the changes broke the tests.

However, after passing a couple of hours trying to understand what was going on (using a debugger, etc.), I figured out that related data structure were not so similar and maybe the shortcut of #2978 cannot really be applied this way.

I don't master the code so here, I clearly need help to have a proper fix.

Reply:
I see. Let me talk with some folks to see how we want to approach this. Will get back to you shortly

Reply:
Alright, I figured out how to fix the 3 broken unit tests if we bring this change back:
```
--- a/tests/objects/test_link.py
+++ b/tests/objects/test_link.py
@@ -138,7 +138,7 @@ class TestLink:
         knowledge_base_r = event_loop.run_until_complete(knowledge_svc.get_relationships(dict(edge='has_admin')))
         assert len(knowledge_base_r) == 1
 
-    def test_create_relationship_source_fact(self, event_loop, ability, executor, operation, knowledge_svc, fire_event_mock):
+    def test_create_relationship_source_fact(self, event_loop, ability, executor, operation, data_svc, knowledge_svc, fire_event_mock):
         test_executor = executor(name='psh', platform='windows')
         test_ability = ability(ability_id='123', executors=[test_executor])
         fact1 = Fact(trait='remote.host.fqdn', value='dc')
@@ -149,6 +149,7 @@ class TestLink:
                               adversary=Adversary(name='sample', adversary_id='XYZ', atomic_ordering=[],
                                                   description='test'),
                               source=Source(id='test-source', facts=[fact1]))
+        event_loop.run_until_complete(data_svc.store(operation.source))
         event_loop.run_until_complete(operation._init_source())
         event_loop.run_until_complete(link1.create_relationships([relationship], operation))
 
@@ -161,7 +162,7 @@ class TestLink:
         assert len(fact_store_operation) == 1
         assert len(fact_store_operation_source[0].collected_by) == 2
 
-    def test_save_discover_seeded_fact_not_in_command(self, event_loop, ability, executor, operation, knowledge_svc, fire_event_mock):
+    def test_save_discover_seeded_fact_not_in_command(self, event_loop, ability, executor, operation, knowledge_svc, data_svc, fire_event_mock):
         test_executor = executor(name='psh', platform='windows')
         test_ability = ability(ability_id='123', executors=[test_executor])
         fact1 = Fact(trait='remote.host.fqdn', value='dc')
@@ -172,6 +173,7 @@ class TestLink:
                               adversary=Adversary(name='sample', adversary_id='XYZ', atomic_ordering=[],
                                                   description='test'),
                               source=Source(id='test-source', facts=[fact1, fact2]))
+        event_loop.run_until_complete(data_svc.store(operation.source))
         event_loop.run_until_complete(operation._init_source())
         event_loop.run_until_complete(link.save_fact(operation, fact2, 1, relationship))
 
diff --git a/tests/objects/test_operation.py b/tests/objects/test_operation.py
index d3d4ecc..be95d9d 100644
--- a/tests/objects/test_operation.py
+++ b/tests/objects/test_operation.py
@@ -427,6 +427,7 @@ class TestOperation:
 
     def test_facts(self, event_loop, app_svc, contact_svc, file_svc, data_svc, learning_svc, fire_event_mock,
                    op_with_learning_and_seeded, make_test_link, make_test_result, knowledge_svc):
+        event_loop.run_until_complete(data_svc.store(op_with_learning_and_seeded.source))
         test_link = make_test_link(9876)
         op_with_learning_and_seeded.add_link(test_link)
```

Reply:
Those tests in particular need to manually store the operation fact sources in the data service - the data service loads fact sources on server startup, so I essentially had to replicate that behavior with the dummy fact sources for the unit test

Reply:
@uruwhy, thank you 👍 
I applied your changes and all tests are fixed now.

For the CI/CD to work properly again, only one thing is missing:  **The missing value of the `SONAR_TOKEN` environment variable** (see [my previous comment](#issuecomment-2211835721)).

Reply:
I'm not sure i'm seeing the same sonar issue that you are, but I did also see some flake8 issues from the latest runs:
```
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

app/api/v2/managers/fact_source_manager.py:3:1: E302 expected 2 blank lines, found 1
app/service/data_svc.py:484:1: W293 blank line contains whitespace
```

Reply:
@uruwhy, thanks, I missed them. Fixed.

Now the `SONAR_TOKEN` missing value appears in the _SonarCloud Scan_ task:

> Set the SONAR_TOKEN env variable.

`.github/workflows/quality.yml:64`:

```yaml
      - name: SonarCloud Scan
        uses: SonarSource/sonarcloud-github-action@49e6cd3b187936a73b8280d59ffd9da69df63ec9
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}  # Needed to get PR information, if any
          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
```

Reply:
So PRs from forks don't have access to our github secrets, and we'd have to make some major workflow adjustments to accommodate this sort of thing. However, I looked through the coverage.xml manually and verified that all your new lines of code are covered